design/design.md

```diff
@@ -0,0 +1,112 @@
+As an optimization of the Telescope with prehashing, repetitions and bounded
+DFS, we could even eliminate this integer from the certificate if the
+number of allowed repetitions r = 1.
+Finally, the simple lottery, while having large proof sizes, offers very
+fast proving times which might be important for some applications.
+For the decentralized setting and weighted settings, we again have four
+possible algorithms:
+the three Telescope algorithms combined with the simple lottery, and the
+simple lottery itself.
+I propose we implement Telescope with prehashing, repetitions and bounded DFS,
+as well as, the single simple lottery.
+Similar to above, the other Telescope schemes offer little advantage over this
+Telescope algorithm.
+The simple lottery, however, offers the smallest communication complexity mu
+and fastest aggregation time.
+It can likely be the top choice for applications where certificate size is not
+important.
+We should decide if we want to provide both unweighted and weighted versions, or
+simply the weighted versions.
+At least for the centralized setting, unweighted versions can be concretely more
+efficient due to no lottery and less complexity.
+For the decentralized setting, an unweighted version could be more efficient as
+well due to not needing to sample the binomial distribution.
+
+What hash function to use for ALBA?
+Ideally, the user should be able to provide his own hash function.
+Hash functions evolve, old ones get broken (sha1) and so our ALBA implementation
+shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, providing the right interfaces should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+We might think about offering a default hash function (e.g. blake2), but this
+should be done with care.
+As many applications use a hash function for multiple sub-protocols, it is
+important to specify distinct hash function seeds for each so that inputs to
+the hash function do not collide.
+
+How mathematically rigorous should our implementation be?
+At this point, probably not very rigorous.
+We need to roll out a prototype so that our primary clients (mithril, etc) can
+incorporate it in their prototypes, and also so that the external community can
+start playing with it as well.
+This means that tiny errors when converting a uniformly random hash output to
+an ALBA specific distribution will be ignored, and sampling a binomial / poisson
+distribution for the weighted ALBA will follow a simple but not well understood
+algorithm.
+
+How will parameters be chosen / computed?
+I propose that we separate parameter derivation for the actual ALBA code.
+This means that converting user friendly parameters such as n_p, n_f, lambdas to
```

Comment:
Why is this so? From a library user's perspective, $d$, $q$ and the like are opaque and internal to the ALBA scheme. The $n_p$ $n_f$ and $\lambda$ parameters are pervasive in the original paper and seem to be an integral part of the public API of ALBA.

Reply:
To ensure backward compability, you may need to provide d,q and co..